### PR TITLE
Add OfficeToPdf connector

### PR DESCRIPTION
**Pull request by pierre-yves-monnet**

### CODE REVIEW

**Comment by pierre-yves-monnet** [comment]
Hello @camunda-community-hub/devrel, here a new PR for this documentation page

File: README.md [hunk 1]
```diff
@@ -50,9 +50,12 @@ Coming Soon!
...
+| Connector                 | JAR with dependencies                                                                 | Code repository                                                                   | Documentation                                                                                         | License | Lifecycle |
+|---------------------------|---------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------| ------- | --------- |
+| Office (Word, OpenOffice) | [Community] (https://github.com/camunda-community-hub/zeebe-cherry-officepdf-workers) | [github](https://github.com/camunda-community-hub/zeebe-cherry-officepdf-workers) | [ReadMe](https://github.com/camunda-community-hub/zeebe-cherry-officepdf-workers/blob/main/README.md) |     [Apache 2.0]     |    [![](https://img.shields.io/badge/Lifecycle-Stable-brightgreen)](https://github.com/Camunda-Community-Hub/community/blob/main/extension-lifecycle.md#stable-)|
```
Comment (MaxTru) [inline README.md:55]:
```suggestion
| Office (Word, OpenOffice) | [Community](https://github.com/camunda-community-hub/zeebe-cherry-officepdf-workers) | [github](https://github.com/camunda-community-hub/zeebe-cherry-officepdf-workers) | [ReadMe](https://github.com/camunda-community-hub/zeebe-cherry-officepdf-workers/blob/main/README.md) |     [Apache 2.0]     |    [![](https://img.shields.io/badge/Lifecycle-Stable-brightgreen)](https://github.com/Camunda-Community-Hub/community/blob/main/extension-lifecycle.md#stable-)|
```

File: README.md [hunk 2]
```diff
@@ -50,9 +50,12 @@ Coming Soon!
...
+| PDF                       | [Community] (https://github.com/camunda-community-hub/zeebe-cherry-officepdf-workers) | [github](https://github.com/camunda-community-hub/zeebe-cherry-officepdf-workers) | [ReadMe](https://github.com/camunda-community-hub/zeebe-cherry-officepdf-workers/blob/main/README.md) |     [Apache 2.0]     |    [![](https://img.shields.io/badge/Lifecycle-Stable-brightgreen)](https://github.com/Camunda-Community-Hub/community/blob/main/extension-lifecycle.md#stable-)|
```
Comment (MaxTru) [inline README.md:56]:
```suggestion
| PDF                       | [Community](https://github.com/camunda-community-hub/zeebe-cherry-officepdf-workers) | [github](https://github.com/camunda-community-hub/zeebe-cherry-officepdf-workers) | [ReadMe](https://github.com/camunda-community-hub/zeebe-cherry-officepdf-workers/blob/main/README.md) |     [Apache 2.0]     |    [![](https://img.shields.io/badge/Lifecycle-Stable-brightgreen)](https://github.com/Camunda-Community-Hub/community/blob/main/extension-lifecycle.md#stable-)|
```

File: README.md [hunk 3]
```diff
@@ -50,9 +50,12 @@ Coming Soon!
...
+| BPMN Operation (message) | [Community] (https://github.com/camunda-community-hub/zeebe-cherry-bpmn-operations) | [github](https://github.com/camunda-community-hub/zeebe-cherry-officepdf-workers) | [ReadMe](https://github.com/camunda-community-hub/zeebe-cherry-bpmn-operations/blob/main/README.md) |     [Apache 2.0]     |    [![](https://img.shields.io/badge/Lifecycle-Stable-brightgreen)](https://github.com/Camunda-Community-Hub/community/blob/main/extension-lifecycle.md#stable-)|
```
Comment (MaxTru) [inline README.md:57]:
```suggestion
| BPMN Operation (message) | [Community](https://github.com/camunda-community-hub/zeebe-cherry-bpmn-operations) | [github](https://github.com/camunda-community-hub/zeebe-cherry-officepdf-workers) | [ReadMe](https://github.com/camunda-community-hub/zeebe-cherry-bpmn-operations/blob/main/README.md) |     [Apache 2.0]     |    [![](https://img.shields.io/badge/Lifecycle-Stable-brightgreen)](https://github.com/Camunda-Community-Hub/community/blob/main/extension-lifecycle.md#stable-)|
```

Comment (MaxTru) [inline README.md:57]: This Connector Repo misses an element template. Am I missing something?

**Review by MaxTru — DISMISSED** [review]
Really nice to see first contributions to this!

Can you please take a look at my comments?

**Comment by tmetzke** [comment]
Hi @pierre-yves-monnet, thanks for those contributions! There are some things I would like to see sorted out before we add this to the list:

* Is there a straightforward guide on how people can use this as a Connector? What is the Connector Function to use? There is no `OutboundConnectorFunction` implementation I can see in the repositories. Do people have to use the Cherry Framework somehow as well here? It would be nice to create a clear path for Connector users here.
* Are there any plans for creating releases for those repositories? Currently, there are none and I believe the projects will be easier to consume if there are. Connector users can come from various backgrounds and prefer different distribution channels. For example, offering the projects as Maven Central artifacts (something the Community Hub readily provides, if I am not mistaken) can be very helpful.

Overall, looking at the code, I don't have the feeling that this is really Connector-focused yet. I rather have the impression that these are general (and very helpful) JobWorker implementations. Happy to discuss 🙂

**Review by tmetzke — DISMISSED** [review]
Hi @pierre-yves-monnet, thanks for those contributions! There are some things I would like to see sorted out before we add this to the list:

* Is there a straightforward guide on how people can use this as a Connector? What is the Connector Function to use? There is no `OutboundConnectorFunction` implementation I can see in the repositories. Do people have to use the Cherry Framework somehow as well here? It would be nice to create a clear path for Connector users here.
* Are there any plans for creating releases for those repositories? Currently, there are none and I believe the projects will be easier to consume if there are. Connector users can come from various backgrounds and prefer different distribution channels. For example, offering the projects as Maven Central artifacts (something the Community Hub readily provides, if I am not mistaken) can be very helpful.

Overall, looking at the code, I don't have the feeling that this is really Connector-focused yet. I rather have the impression that these are general (and very helpful) JobWorker implementations. Happy to discuss 🙂

**Comment by berndruecker** [comment]
@pierre-yves-monnet Might be worth to check the difference between a connector and a worker, see also https://camunda.com/blog/2022/10/camundacon-2022-qa-bernd-daniel/ --> 

_What is the difference between a Job Worker and a Connector?

The difference is actually not too big. Especially the code is pretty similar. But the connector works on a slightly higher abstraction. For example, you cannot simply access all process variables, but need to explicitly define what data will be available. Additionally, you will have support to handle secrets properly, which you have to handle on your own in a Job Worker.

This reduces the scope of what a connector can do, which allows us to build different runtimes. So while in Spring Boot the difference is hard to see, you can run a connector function in SaaS in a stateless Google Function, which is not the case for a Job Worker.

Anyway, I expect that we will add some more syntactic sugar on connectors over time, so this might even get the default way of waiting Java glue code in future, making it unnecessary to leverage the “low level” Job Worker API. But honestly, this also depends on how the community is picking things up._

And I just recognized, I forgot to talk about element templates there 😨

**Comment by pierre-yves-monnet** [comment]
Hello, Thank you for your comment.
If I want to move this code to a pure Connector, what is the path then?

I understand (correct me if I'm wrong)

1/ one repository per connector (or can we group multiple connectors in the same repository?)

2/ the code should be purely "extends AbstractConnector" 

3/ using the https://github.com/camunda-community-hub/community-action-maven-release, I will create a release and push it to the Maven Central.

4/ @berndruecker: the element template is generated by the Cherry Framework. I can save it in the repository (under an "element-template" path?).

**Comment by berndruecker** [comment]
Great questions - we haven't yet settled on the exact answers to your question.

**Comment by tmetzke** [comment]
Hi @pierre-yves-monnet, good questions indeed 👍 

I agree with what Bernd added here so far.

Regarding 2/, one Connector is ideally implementing the `OutboundConnectorFunction` interface, yes. Have a look at [our template](https://github.com/camunda/connector-template) as well for reference.

Regarding 3/, we currently offer two ways of consuming our out-of-the-box Connectors:
1. An uber JAR that contains all necessary dependencies of the Connector.
1. A regular thin JAR that only contains the Connector's code.
The uber JAR can easily be added to [Docker images](https://github.com/camunda/connectors-bundle/blob/main/Dockerfile) or the [pre-packaged runtime](https://github.com/camunda/connector-sdk/tree/main/runtime-job-worker#automatic-connector-discovery). The thin JAR can easily be added as a Maven/Gradle dependency and be used in [Spring Zeebe](https://github.com/camunda-community-hub/spring-zeebe#run-outboundconnectors), for example.

Regarding 4/, we also provide the element template with a GitHub release of every Connector. That way, it's easy to see for users which element template is associated with which release of your Connector behavior.

I agree that we should create some guidelines on this, @berndruecker 👍

**Comment by pierre-yves-monnet** [comment]
Hello,
Great, I will see how to move these objects to connectors. In the first move, I will use the same GitHub: these if 10 connectors for the moment CMIS collection will add 8 more.

Is that acceptable that the connector implements OutboundConnectorFunction  AND extends a class?
For the Cherry dashboard, we want the developer to give more information on input/ output in order to generate the template modeler and display information to the administrator and designer (https://github.com/camunda-community-hub/zeebe-cherry-framework).  Our target is to be 100% compliant with the Connector framework and adds more information.

PS: Actually, all connector extends AbstrraceConnector, which implement OutboundConnectorFunction , but we can move the implementation one level up.


public abstract class AbstractConnector extends AbstractRunner implements OutboundConnectorFunction {

**Comment by tmetzke** [comment]
Hi @pierre-yves-monnet, excited to see those transformations taking place. Thanks for putting in the effort here 👍 

As far as I can see, it should be enough that the Connector functions technically implement the interface. We use a `ServiceLoader` to load all classes that implement the interface, which should work regardless of intermediate classes in the hierarchy.

File: README.md (hunk 3 above)
Comment (pierre-yves-monnet) [reply README.md]: The connector is integrated into the Cherry Framework. The framework generates automatically the element-template. But I add it for you in a element-templates folder.

**Comment by pierre-yves-monnet** [comment]
Hello,

To follow your direction, I do the job for one connector. If you are OK with this contribution, I will do the same job for the other connectors. Each connector will have its GitHub repository.

1/ I created one repository per connector. The first contribution is the SendBPMNMessage (https://github.com/camunda-community-hub/camunda-8-connector-sendBPMNmessage)

2/ This connector implements the interface OutboundConnectorFunction
`public class SendBPMNMessageFunction extends AbstractConnector implements OutboundConnectorFunction {`

It extends the AbstractConnector from the Cherry Framework to get all advantages on this framework. If you don't like this, I can go one step ahead, keep this class without the extension, and creates a class in the Cherry Collection that extends this one - but this means more work)

3/ You will find a folder `element-templates` with the element templates. 

Thank you for your review,

**Comment by tmetzke** [comment]
Thanks, @pierre-yves-monnet, for adjusting this 👍 
I will give your project a closer look as soon as I can (probably tomorrow, we just had a release to finish).
In the meantime, just wanted to check with you, are you certain about the lifecycle banner `stable`? If you're OK with the guarantees and expectations users have with this, that's OK, just wanted to make sure this was intentional. 🙂

**Comment by berndruecker** [comment]
Lifecycle badges are explained here: https://github.com/Camunda-Community-Hub/community/blob/main/extension-lifecycle.md. I would spontaneously see this as POC or Incubating

**Comment by tmetzke** [comment]
Hi @pierre-yves-monnet, thanks again for adjusting your Connector 👍 
There are a couple of things to look at and discuss, probably.

## Included runtime environments

A Connector should come without any runtime environment since it is supposed to be environment-agnostic. Your code however comes with Spring Zeebe and the `zeebe-cherry-framework`, which in turn brings along the Zeebe Java client as well as the Connector Runtime `0.2.x`.

**Why is this an issue?**
If someone wants to embed your Connector in Spring Zeebe or any other runtime environment (either as a Maven dependency or as a fat JAR including dependencies) this will become an issue. There will be multiple runtimes available that can execute Connectors which will most probably lead to conflicts.

**What can you do about it?**
Remove all runtime-related dependencies. 

Spring Zeebe can simply be removed, as people can use Spring Zeebe directly to execute Connectors (see [documentation](https://github.com/camunda-community-hub/spring-zeebe/#run-outboundconnectors)). No need to include it in your project, I believe.

The cherry framework dependency could be a bit more tricky. I don't have much knowledge about it, but it seems to bring along a runtime (Connector Runtime `0.2.x`) - this is counterproductive. Either you can remove the dependency completely or create different artifacts in that framework, one of them covering the basics you need here in a Connector without bringing along a runtime.

Please also try to avoid adjusting versions of something like the Jakarta Validation API, since this is determined by the SDK's `connector-validation` module and can lead to version conflicts quite quickly when people use your Connector together with other Connectors that use that module as well.

## SDK version `0.2.x`

Please try to move to Connector SDK `0.3.0`. Have a look at the updated [Connector template repository](https://github.com/camunda/connector-template) for inspiration. The POM in the repository also contains the capabilities to build a fat JAR automatically.  and also the [Update Guide](https://docs.camunda.io/docs/next/guides/update-guide/connectors/020-to-030/). 



Let me know what you think.

**Comment by tmetzke** [comment]
Something I noticed just now:

* You are using Spring to inject a `ZeebeContainer` for example (also the main function is annotated as `@Component`). This makes the Connector less reusable since it can only work in Spring environments. However, not all Connector runtimes will be in that realm. So, I would urge you to remove Spring-related stuff from the Connector.
* Connectors don't throw `ZeebeBpmnError`s, they throw technical errors and can provide error code with them. This is new in Connector SDK `0.3.0`, you can read about it in the example over at https://docs.camunda.io/docs/next/components/connectors/custom-built-connectors/connector-sdk/#runtime-logic.

**Comment by pierre-yves-monnet** [comment]
Hello,
Connector SendBPMNMessage comes too early because it needs to have a ZeebeClient, which is not available in the ConnectorSDK at this moment.

So, I focus on the first connector from the Office Collection, and I choose the OfficeToPdf connector.
Thank you for reviewing it.
Best

**Comment by tmetzke** [comment]
> Hello, Connector SendBPMNMessage comes too early because it needs to have a ZeebeClient, which is not available in the ConnectorSDK at this moment.

As mentioned in the [SDK issue](https://github.com/camunda/connector-sdk/issues/273), a Connector should not rely on the Zeebe client being available. If you want to connect to a C8 cluster, create a Zeebe client in your Connector as you would with creating a client to any other external system.

**Comment by igpetrov** [comment]
Hi @pierre-yves-monnet,

Thank you for your contribution.

Few thoughts on reviewing the suggested connector:
- Could you please declare the license?
- Could you please add an instruction for the users on how to use it?
- I noticed that you're trying to publish your connector but [it won't pass](https://github.com/camunda-community-hub/camunda-8-connector-officetopdf/actions/runs/3662182852/jobs/6191087634). Is it something on your radar?

Thank you.

**Comment by pierre-yves-monnet** [comment]
Hello @igpetrov

Could you please declare the license?
==> Done.

Could you please add an instruction for the users on how to use it?
==> I add them in the README. I hope this is clear.
I noticed that you're trying to publish your connector, but [it won't pass](https://github.com/camunda-community-hub/camunda-8-connector-officetopdf/actions/runs/3662182852/jobs/6191087634). Is it something on your radar?
==> Yes, I got some issue understanding what was missing, but now it's good; publication is OK.

**Review by igpetrov — APPROVED** [review]
LGTM. Thank you for the provided changes!

**Review by chillleader — APPROVED** [review]

**Pull request merged**